build-android/app/src/main/java/com/google/firebase/codelab/friendlychat/MessageItemAdapter.kt

```diff
@@ -15,30 +15,43 @@
  */
 package com.google.firebase.codelab.friendlychat
 
+import android.graphics.Color
 import android.util.Log
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
 import android.widget.ImageView
 import android.widget.TextView
+import androidx.constraintlayout.widget.ConstraintLayout
+import androidx.core.content.ContextCompat
 import androidx.recyclerview.widget.RecyclerView
 import com.bumptech.glide.Glide
 import com.firebase.ui.database.FirebaseRecyclerAdapter
 import com.firebase.ui.database.FirebaseRecyclerOptions
+import com.google.firebase.codelab.friendlychat.MainActivity.Companion.ANONYMOUS
+import com.google.firebase.codelab.friendlychat.databinding.ItemMessageBinding
 import com.google.firebase.codelab.friendlychat.model.FriendlyMessage
 import com.google.firebase.ktx.Firebase
 import com.google.firebase.storage.ktx.storage
-import de.hdodenhof.circleimageview.CircleImageView
 
 // The FirebaseRecyclerAdapter class and options come from the FirebaseUI library
 // See: https://github.com/firebase/FirebaseUI-Android
-class MessageItemAdapter(options: FirebaseRecyclerOptions<FriendlyMessage>) :
+class MessageItemAdapter(
+    options: FirebaseRecyclerOptions<FriendlyMessage>,
+    currentUserName: String?
+) :
     FirebaseRecyclerAdapter<FriendlyMessage, MessageItemAdapter.MessageItemViewHolder>(options) {
 
+    private val currentUserName = currentUserName
+
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MessageItemViewHolder {
```

Comment:
Have you considered using multiple different `viewTypes` so you don't have to do as much UI logic in code?

If you override [`getItemViewType`](https://developer.android.com/reference/androidx/recyclerview/widget/RecyclerView.Adapter#getItemViewType(int)) you can choose a type for each item. Maybe image messages and text messages could be different types? And then in `onCreateViewHolder` you'd get the `viewType` argument which would let you pick the `R.layout` resource that you want. 

Might make using the viewBinding a bit tricky ... this is totally optional just wondering if you considered it?

Reply:
I considered it when I was trying to display the user-owned messages on the right side of the layout, but decided that idea was too complicated / too distracting for the codelab... but I did not consider it for this use case. I'll try it and see what it looks like, and then once I get to updating the markdown for the codelab, I'll confirm whether it's too complex or not and refactor as needed. Thanks for the suggestion